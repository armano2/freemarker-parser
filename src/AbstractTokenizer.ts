export default abstract class AbstractTokenizer {
  protected template : string = ''
  protected length : number = 0
  protected index : number = 0

  protected charAt (i : number) : string {
    return this.template.charAt(i)
  }

  protected charCodeAt (i : number) : number {
    return this.template.charCodeAt(i)
  }

  protected init (template : string) : void {
    this.template = template
    this.length = template.length
    this.index = 0
  }
}
