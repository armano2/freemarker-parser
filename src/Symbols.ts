import { ETypeSymbol } from './Types'

export interface ISymbol {
  startToken : string
  endToken : string
  type : ETypeSymbol
}

export const symbols : ISymbol[] = [
  { startToken: '</#', endToken: '>', type: ETypeSymbol.DirectiveEnd },
  { startToken: '<#', endToken: '>', type: ETypeSymbol.Directive },
  { startToken: '<@', endToken: '>', type: ETypeSymbol.Macro },
  // tslint:disable-next-line:no-invalid-template-strings
  { startToken: '${', endToken: '}', type: ETypeSymbol.Interpolation },
]

export const whitespaces : string[] = [
  ' ',
  '\t',
  '\n',
  '\r',
]
