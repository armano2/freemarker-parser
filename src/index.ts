import { Parser } from './Parser'

export {
  Parser,
}
