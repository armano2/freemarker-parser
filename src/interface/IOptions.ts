export interface IOptions {
  squareTags?: boolean;
  // squareInterpolation? : boolean // TODO: unsupported
  // hashInterpolation? : boolean // TODO: unsupported
  parseLocation?: boolean;
  // useCamelCaseNaming? : boolean // TODO: unsupported
}
