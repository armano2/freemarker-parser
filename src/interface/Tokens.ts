import { ENodeType } from '../Symbols'
import { ISourceLocation } from './ISourceLocation'

export interface ILoc {
  start : number
  end : number,
  loc? : {
    start : ISourceLocation,
    end : ISourceLocation,
  }
}

export interface IToken extends ILoc {
  type : ENodeType
  startTag? : string
  endTag? : string
  params? : string
  text : string
}
