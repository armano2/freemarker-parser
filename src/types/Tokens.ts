import NodeNames from '../enum/NodeNames'
import { ENodeType } from '../Symbols'

export interface IDirectivesTypes {
  [n : string] : NodeNames
}

export interface IToken extends ILoc {
  type : ENodeType
  startTag? : string
  endTag? : string
  params? : string
  text : string
  isClose : boolean
}

export interface ILoc {
  start : number
  end : number
}
