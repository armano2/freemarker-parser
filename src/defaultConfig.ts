import { Options } from './interface/Options';

const defaultConfig: Options = {
  squareTags: false,
  parseLocation: true,
};

export default defaultConfig;
