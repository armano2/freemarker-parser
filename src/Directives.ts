import NodeNames from './enum/NodeNames'
import { IDirectivesTypes } from './types/Tokens'

const Directives : IDirectivesTypes = {
  if: NodeNames.Condition,
  else: NodeNames.Else,
  elseif: NodeNames.ConditionElse,
  list: NodeNames.List,
  include: NodeNames.Include,
  assign: NodeNames.Assign,
  attempt: NodeNames.Attempt,
  compress: NodeNames.Compress,
  // escape: NodeNames.Escape,
  // noescape: NodeNames.NoEscape,
  // fallback: NodeNames.Fallback,
  function: NodeNames.Function,
  // flush: NodeNames.Flush,
  global: NodeNames.Global,
  import: NodeNames.Import,
  local: NodeNames.Local,
  // lt: NodeNames.Lt,
  macro: NodeNames.Macro,
  // nested: NodeNames.Nested,
  // nt: NodeNames.Nt,
  recover: NodeNames.Recover,
  // recurse: NodeNames.Recurse,
  return: NodeNames.Return,
  // rt: NodeNames.Rt,
  // setting: NodeNames.Setting,
  // stop: NodeNames.Stop,
  switch: NodeNames.Switch,
  case: NodeNames.SwitchCase,
  default: NodeNames.SwitchDefault,
  break: NodeNames.Break,
  // t: NodeNames.T,
  // visit: NodeNames.Visit,
  noparse: NodeNames.Text,
  noParse: NodeNames.Text,
}

export default Directives
