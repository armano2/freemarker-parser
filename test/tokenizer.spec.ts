import { Tokenizer } from '../src'
import { IToken } from '../src/interface/Tokens'
import { ENodeType } from '../src/Symbols'
import Tester from './testers/TokenizerTester'

const tokenizer = new Tokenizer()

function parse (text : string) : IToken[] {
  return tokenizer.parse(text)
}

describe('parsing directives', () => {
  it('no arguments', () => {
    const tokens = parse('<#foo>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false)
  })
  it('no arguments, self closing', () => {
    const tokens = parse('<#foo/>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false)
  })
  it('many, no arguments', () => {
    const tokens = parse('<#foo><#foo>')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false)
  })
  it('many, no arguments, with text', () => {
    const tokens = parse('foo<#foo>foo<#foo>foo')
    Tester.instance(tokens)
      .hasCount(5)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
  })
  it('no arguments, with close tag', () => {
    const tokens = parse('<#foo></#foo>')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(true)
  })
  it('with arguments', () => {
    const tokens = parse('<#foo bar>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Directive).hasParams('bar').isCloseTag(false)
  })
  it('many, with arguments', () => {
    const tokens = parse('<#foo bar><#foo bar test><#foo bar less>')
    Tester.instance(tokens)
      .hasCount(3)
      .isType(ENodeType.Directive).hasParams('bar').isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Directive).hasParams('bar test').isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Directive).hasParams('bar less').isCloseTag(false)
  })
})

describe('parsing macros', () => {
  it('no arguments', () => {
    const tokens = parse('<@foo>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
  })
  it('no arguments, self closing', () => {
    const tokens = parse('<@foo/>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
  })
  it('many, no arguments', () => {
    const tokens = parse('<@foo><@foo>')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
  })
  it('many, no arguments, with text', () => {
    const tokens = parse('foo<@foo>foo<@foo>foo')
    Tester.instance(tokens)
      .hasCount(5)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
  })
  it('no arguments, with close tag', () => {
    const tokens = parse('<@foo></@foo>')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Macro).hasParams(undefined).isCloseTag(true)
  })
  it('with arguments', () => {
    const tokens = parse('<@foo bar>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Macro).hasParams('bar').isCloseTag(false)
  })
  it('many, with arguments', () => {
    const tokens = parse('<@foo bar><@foo bar test><@foo bar less>')
    Tester.instance(tokens)
      .hasCount(3)
      .isType(ENodeType.Macro).hasParams('bar').isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Macro).hasParams('bar test').isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Macro).hasParams('bar less').isCloseTag(false)
  })
})

describe('parsing comments', () => {
  it('coment with text', () => {
    const tokens = parse('<#--  <@d></@d>  -->')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Comment).hasNoParams().isCloseTag(false).hasText('  <@d></@d>  ')
  })
  it('coment in directive', () => {
    const tokens = parse('<#foo><#--  foo  --></#foo>')
    Tester.instance(tokens)
      .hasCount(3)
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Comment).hasNoParams().isCloseTag(false).hasText('  foo  ')
      .nextToken()
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(true)
  })
})

describe('parsing text', () => {
  it('empty text', () => {
    const tokens = parse('')
    Tester.instance(tokens)
      .hasCount(0)
  })
  it('raw text', () => {
    const tokens = parse('<foo>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('<foo>')
  })
  it('text after directive', () => {
    const tokens = parse('<#foo>foo')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
      .nextToken()
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
  })
  it('text before directive', () => {
    const tokens = parse('foo<#foo>')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Directive).hasParams(undefined).isCloseTag(false)
  })
})

describe('html', () => {
  it('simple', () => {
    const tokens = parse('<p>This is include-subdir.ftl</p>')
    Tester.instance(tokens)
      .hasCount(1)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('<p>This is include-subdir.ftl</p>')
  })
  it('advance', () => {
    const tokens = parse('<p>This is include-subdir.ftl</p><#include "include-subdir2.ftl">')
    Tester.instance(tokens)
      .hasCount(2)
      .isType(ENodeType.Text).hasNoParams().isCloseTag(false).hasText('<p>This is include-subdir.ftl</p>')
      .nextToken()
      .isType(ENodeType.Directive).hasParams('"include-subdir2.ftl"').isCloseTag(false)
  })
})

describe('names', () => {
  it('directive', () => {
    const tokens = parse('<#foo/><#Foo/><#FoO/><#Fo_O/>')
    Tester.instance(tokens)
      .hasCount(4)
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false).hasText('foo')
      .nextToken()
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false).hasText('Foo')
      .nextToken()
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false).hasText('FoO')
      .nextToken()
      .isType(ENodeType.Directive).hasNoParams().isCloseTag(false).hasText('Fo_O')
  })
})
