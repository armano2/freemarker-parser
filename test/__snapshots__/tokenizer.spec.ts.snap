// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html advance 1`] = `
Array [
  Object {
    "end": 33,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "<p>This is include-subdir.ftl</p>",
    "type": "Text",
  },
  Object {
    "end": 65,
    "endTag": ">",
    "params": "\\"include-subdir2.ftl\\"",
    "start": 33,
    "startTag": "<#",
    "text": "include",
    "type": "OpenDirective",
  },
]
`;

exports[`html simple 1`] = `
Array [
  Object {
    "end": 33,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "<p>This is include-subdir.ftl</p>",
    "type": "Text",
  },
]
`;

exports[`names directive 1`] = `
Array [
  Object {
    "end": 7,
    "endTag": "/>",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 14,
    "endTag": "/>",
    "params": undefined,
    "start": 7,
    "startTag": "<#",
    "text": "Foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 21,
    "endTag": "/>",
    "params": undefined,
    "start": 14,
    "startTag": "<#",
    "text": "FoO",
    "type": "OpenDirective",
  },
  Object {
    "end": 29,
    "endTag": "/>",
    "params": undefined,
    "start": 21,
    "startTag": "<#",
    "text": "Fo_O",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing comments coment in directive 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 20,
    "endTag": "-->",
    "params": undefined,
    "start": 6,
    "startTag": "<#--",
    "text": "  foo  ",
    "type": "Comment",
  },
  Object {
    "end": 27,
    "endTag": ">",
    "params": undefined,
    "start": 20,
    "startTag": "</#",
    "text": "foo",
    "type": "CloseDirective",
  },
]
`;

exports[`parsing comments coment with text 1`] = `
Array [
  Object {
    "end": 20,
    "endTag": "-->",
    "params": undefined,
    "start": 0,
    "startTag": "<#--",
    "text": "  <@d></@d>  ",
    "type": "Comment",
  },
]
`;

exports[`parsing directives many, no arguments 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 12,
    "endTag": ">",
    "params": undefined,
    "start": 6,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing directives many, no arguments, with text 1`] = `
Array [
  Object {
    "end": 3,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
  Object {
    "end": 9,
    "endTag": ">",
    "params": undefined,
    "start": 3,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 12,
    "endTag": undefined,
    "params": undefined,
    "start": 9,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
  Object {
    "end": 18,
    "endTag": ">",
    "params": undefined,
    "start": 12,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 21,
    "endTag": undefined,
    "params": undefined,
    "start": 18,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
]
`;

exports[`parsing directives many, with arguments 1`] = `
Array [
  Object {
    "end": 10,
    "endTag": ">",
    "params": "bar",
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 25,
    "endTag": ">",
    "params": "bar test",
    "start": 10,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 40,
    "endTag": ">",
    "params": "bar less",
    "start": 25,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing directives no arguments 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing directives no arguments, self-closing 1`] = `
Array [
  Object {
    "end": 7,
    "endTag": "/>",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing directives no arguments, with close tag 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 13,
    "endTag": ">",
    "params": undefined,
    "start": 6,
    "startTag": "</#",
    "text": "foo",
    "type": "CloseDirective",
  },
]
`;

exports[`parsing directives with arguments 1`] = `
Array [
  Object {
    "end": 10,
    "endTag": ">",
    "params": "bar",
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;

exports[`parsing macros many, no arguments 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 12,
    "endTag": ">",
    "params": undefined,
    "start": 6,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
]
`;

exports[`parsing macros many, no arguments, with text 1`] = `
Array [
  Object {
    "end": 3,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
  Object {
    "end": 9,
    "endTag": ">",
    "params": undefined,
    "start": 3,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 12,
    "endTag": undefined,
    "params": undefined,
    "start": 9,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
  Object {
    "end": 18,
    "endTag": ">",
    "params": undefined,
    "start": 12,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 21,
    "endTag": undefined,
    "params": undefined,
    "start": 18,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
]
`;

exports[`parsing macros many, with arguments 1`] = `
Array [
  Object {
    "end": 10,
    "endTag": ">",
    "params": "bar",
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 25,
    "endTag": ">",
    "params": "bar test",
    "start": 10,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 40,
    "endTag": ">",
    "params": "bar less",
    "start": 25,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
]
`;

exports[`parsing macros no arguments 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
]
`;

exports[`parsing macros no arguments, self-closing 1`] = `
Array [
  Object {
    "end": 7,
    "endTag": "/>",
    "params": undefined,
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
]
`;

exports[`parsing macros no arguments, with close tag 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
  Object {
    "end": 13,
    "endTag": ">",
    "params": undefined,
    "start": 6,
    "startTag": "</@",
    "text": "foo",
    "type": "CloseMacro",
  },
]
`;

exports[`parsing macros with arguments 1`] = `
Array [
  Object {
    "end": 10,
    "endTag": ">",
    "params": "bar",
    "start": 0,
    "startTag": "<@",
    "text": "foo",
    "type": "OpenMacro",
  },
]
`;

exports[`parsing text empty text 1`] = `Array []`;

exports[`parsing text raw text 1`] = `
Array [
  Object {
    "end": 5,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "<foo>",
    "type": "Text",
  },
]
`;

exports[`parsing text text after directive 1`] = `
Array [
  Object {
    "end": 6,
    "endTag": ">",
    "params": undefined,
    "start": 0,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
  Object {
    "end": 9,
    "endTag": undefined,
    "params": undefined,
    "start": 6,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
]
`;

exports[`parsing text text before directive 1`] = `
Array [
  Object {
    "end": 3,
    "endTag": undefined,
    "params": undefined,
    "start": 0,
    "startTag": undefined,
    "text": "foo",
    "type": "Text",
  },
  Object {
    "end": 9,
    "endTag": ">",
    "params": undefined,
    "start": 3,
    "startTag": "<#",
    "text": "foo",
    "type": "OpenDirective",
  },
]
`;
